### DISCUSSION TITLE
Overriding plugin configs from the command line?

### DISCUSSION THREAD

**Discussion by oprypin** [mkdocs/mkdocs#2345] [General]
Each MkDocs plugin has only one, permanent configuration - in the YAML file. There aren't such things as specifying an alternate config at will, or overriding individual values **per invocation** of an MkDocs command.

I think that's a bit unfair, seeing as MkDocs itself has several runtime flags `--quiet`, `--verbose`, `--strict`, `--dirty`, etc. The use cases for a plugin to expose its flags would really be in the same category of usecases as for MkDocs itself.

(As an example of this being done, *pytest* plugins [can expose flags](https://docs.pytest.org/en/stable/example/simple.html#control-skipping-of-tests-according-to-command-line-option) that become part of the top-level command line)

For me in particular, right now I'm developing a linter of sorts, that really needs to be run (and ideally configured) as part of a MkDocs site. But it's really slow, and I want to run it only sometimes, ideally if a particular flag is passed. But there's no way to do it.

Well, one way I though of was through environment variables. `ENABLE_FOO_PLUGIN=1 mkdocs build` or something. A bit ugly, but looks like that's what I'll be doing.

Any other ideas?

**Comment by waylan** [comment]
There are a few different ways to approach this.

We could create some generic method to expose all config settings. Or we could provide a method for specific settings to be specifically exposed.

I'm generally opposed to the first as some settings just don't make sense as command line options. And some simply can't be defined on the command line as they are more complex types.

However, the second option might make sense for some plugins. That specific use case hasn't come up yet (command lien flags for plugins). And I'm not sure how that would look. How would you specify from the commandline that a flag is for s specific plugin? Or would you just give the flag a unique name and then programicaly map it to the plugin's config option? But what if two plugins create a flag with the same name? 

I'm not saying its not resolvable. Just that we would need to work all of those matters out. If someone wanted to provide a full proposal, I would be willing to entertain it. If we can find a workable solution, then a PR would be welcome.

**Comment by oprypin** [comment]
Ooh I've been doing custom handling such as

```yaml
  - my_plugin:
      enabled: false
      enable_on_env: ENABLE_FOO_PLUGIN
```

But now I realize that with https://github.com/mkdocs/mkdocs/pull/2267 this could be just

```yaml
  - my_plugin:
      enabled: !ENV ENABLE_FOO_PLUGIN
```

-- only that's not released yet.

**Comment by pstavirs** [comment]
One possible solution could to override any mkdocs config (either _mkdocs'_ or a _plugin's_) with variable assignments on the command line like `make` does e.g.

```
mkdocs build plugins:with-pdf:verbose=TRUE
```

I've used `:` as a separator character to specify the hierarchy. Some other character could also be used instead.

This can possibly also address #1499
